File: core/src/com/google/zxing/aztec/AztecReader.java
/*
 * Copyright 2010 ZXing authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.zxing.aztec;

import com.google.zxing.BarcodeFormat;
import com.google.zxing.BinaryBitmap;
import com.google.zxing.ChecksumException;
import com.google.zxing.DecodeHintType;
import com.google.zxing.FormatException;
import com.google.zxing.NotFoundException;
import com.google.zxing.Reader;
import com.google.zxing.Result;
import com.google.zxing.ResultMetadataType;
import com.google.zxing.ResultPoint;
import com.google.zxing.ResultPointCallback;
import com.google.zxing.common.DecoderResult;
import com.google.zxing.aztec.decoder.Decoder;
import com.google.zxing.aztec.detector.Detector;

import java.util.Hashtable;

/**
 * This implementation can detect and decode Aztec codes in an image.
 *
 * @author David Olivier
 */
public final class AztecReader implements Reader {

  /**
   * Locates and decodes a Data Matrix code in an image.
   *
   * @return a String representing the content encoded by the Data Matrix code
   * @throws NotFoundException if a Data Matrix code cannot be found
   * @throws FormatException if a Data Matrix code cannot be decoded
   * @throws ChecksumException if error correction fails
   */
  public Result decode(BinaryBitmap image) throws NotFoundException, FormatException {
    return decode(image, null);
  }

  public Result decode(BinaryBitmap image, Hashtable hints)
      throws NotFoundException, FormatException {

    AztecDetectorResult detectorResult = new Detector(image.getBlackMatrix()).detect();
    ResultPoint[] points = detectorResult.getPoints();

    if (hints != null && detectorResult.getPoints() != null) {
      ResultPointCallback rpcb = (ResultPointCallback) hints.get(DecodeHintType.NEED_RESULT_POINT_CALLBACK);
      if (rpcb != null) {
        for (int i = 0; i < detectorResult.getPoints().length; i++) {
          rpcb.foundPossibleResultPoint(detectorResult.getPoints()[i]);
        }
      }
    }

    DecoderResult decoderResult = new Decoder().decode(detectorResult);

    Result result = new Result(decoderResult.getText(), decoderResult.getRawBytes(), points, BarcodeFormat.AZTEC);
    
    if (decoderResult.getByteSegments() != null) {
      result.putMetadata(ResultMetadataType.BYTE_SEGMENTS, decoderResult.getByteSegments());
    }
    if (decoderResult.getECLevel() != null) {
      result.putMetadata(ResultMetadataType.ERROR_CORRECTION_LEVEL, decoderResult.getECLevel().toString());
    }
    
    return result;
  }

  public void reset() {
    // do nothing
  }

}

File: cpp/core/src/zxing/aztec/AztecReader.cpp
// -*- mode:c++; tab-width:2; indent-tabs-mode:nil; c-basic-offset:2 -*-
/*
 *  AztecReader.cpp
 *  zxing
 *
 *  Created by Lukas Stabe on 08/02/2012.
 *  Copyright 2012 ZXing authors All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <zxing/aztec/AztecReader.h>
#include <zxing/aztec/detector/ZxingAztecDetector.h>    // for Detector
#include <vector>                                       // for allocator, vector

#include "zxing/BarcodeFormat.h"                        // for BarcodeFormat::AZTEC_BARCODE
#include "zxing/BinaryBitmap.h"                         // for BinaryBitmap
#include "zxing/DecodeHints.h"                          // for DecodeHints
#include "zxing/Result.h"                               // for Result
#include "zxing/ResultPoint.h"                          // for ResultPoint
#include "zxing/aztec/AztecDetectorResult.h"            // for AztecDetectorResult
#include "zxing/aztec/decoder/Decoder.h"                // for Decoder
#include "zxing/common/BitMatrix.h"                     // for BitMatrix
#include "zxing/common/DecoderResult.h"                 // for DecoderResult
#include "zxing/common/Str.h"                           // for String

namespace pping {
  namespace aztec {
        
    AztecReader::AztecReader() : decoder_() {
      // nothing
    }
        
    FallibleRef<Result> AztecReader::decode(Ref<pping::BinaryBitmap> image) MB_NOEXCEPT_EXCEPT_BADALLOC {
      auto const blackMatrix(image->getBlackMatrix());
      if (!blackMatrix)
          return blackMatrix.error();
      Detector detector(*blackMatrix);
            

      auto const detectorResult(detector.detect());
      if(!detectorResult)
          return detectorResult.error();
            
      std::vector<Ref<ResultPoint> > points(detectorResult->getPoints());
            
      auto const getDecoderResult(decoder_.decode(*detectorResult));
      if(!getDecoderResult)
          return getDecoderResult.error();

      Ref<DecoderResult> decoderResult(*getDecoderResult);

      Ref<Result> result(new Result(decoderResult->getText(),
                                    decoderResult->getRawBytes(),
                                    points,
                                    BarcodeFormat::AZTEC_BARCODE,
                                    decoderResult->getByteSegments()));
            
      return result;
    }
        
    FallibleRef<Result> AztecReader::decode(Ref<BinaryBitmap> image, DecodeHints) MB_NOEXCEPT_EXCEPT_BADALLOC {
      //cout << "decoding with hints not supported for aztec" << "\n" << flush;
      return this->decode(image);
    }
        
    AztecReader::~AztecReader() {
      // nothing
    }
        
    Decoder& AztecReader::getDecoder() {
      return decoder_;
    }
        
  }
}
